File: Day3_Sum_of_Subarray_Minimums/SumOfSubarrayMinimums.py
class Solution(object):
    def sumSubarrayMins(self, arr):
        """
        :type arr: List[int]
        :rtype: int
        """
        M=(10**9+7)
        minis_sum=arr[-1]
        stack=[(arr[-1],arr[-1],len(arr)-1)]

        for i in range(len(arr)-2,-1,-1):
            
            while stack and arr[i]<stack[-1][0]:
                stack.pop()
            if stack:
                #print((stack[-1][2]-i))
                minis_sum+=((stack[-1][2]-i)*(arr[i])+stack[-1][1])
                stack.append((arr[i],(stack[-1][2]-i)*(arr[i])+stack[-1][1],i))
            else:
                minis_sum+=(arr[i]*(len(arr)-i))
                stack.append((arr[i],arr[i]*(len(arr)-i),i))
            #print(i,minis_sum,stack)
        return minis_sum%M


        
        

            
        

File: Day3_Sum_of_Subarray_Minimums/SumOfSubarrayMinimums.java
class Triplet{
    int a,b,c;
    Triplet(int a,int b,int c){
        this.a=a;
        this.b=b;
        this.c=c;
    }
}
class Solution {
    public int sumSubarrayMins(int[] arr) {
        long m=(long)(1e9+7);
        long minis_sum=arr[arr.length-1];
        Stack<Triplet>stack=new Stack<>();
        stack.push(new Triplet(arr[arr.length-1],arr[arr.length-1],arr.length-1));

        for(int i=arr.length-2;i>=0;i--){
            while(!stack.isEmpty() && stack.peek().a>arr[i]){
                stack.pop();
            }
            if(!stack.isEmpty()){
                minis_sum+=((stack.peek().c-i)*arr[i]+stack.peek().b);
                stack.push(new Triplet(arr[i],(stack.peek().c-i)*arr[i]+stack.peek().b,i));
            }
            else{
                minis_sum+=(arr[i]*(arr.length-i));
                stack.push(new Triplet(arr[i],arr[i]*(arr.length-i),i));
            }
        }
        return (int)(minis_sum%m);
        
    }
}
